apps/jobs/views.py

```diff
@@ -2073,3 +2077,160 @@ def get_github_badge_data(
+    for participant_email in all_participants_email:
+        if participant_email in challenge.banned_email_ids:
+            message = "You're a part of {} team and it has been banned from this challenge. \
+            Please contact the challenge host.".format(
+                participant_team.team_name
+            )
+            response_data = {"error": message}
+            return Response(response_data, status=status.HTTP_403_FORBIDDEN)
+
+    # Fetch the number of submissions under progress.
+    submissions_in_progress_status = [
+        Submission.SUBMITTED,
+        Submission.SUBMITTING,
+        Submission.RUNNING,
+    ]
+    submissions_in_progress = Submission.objects.filter(
+        participant_team=participant_team_id,
+        challenge_phase=challenge_phase,
+        status__in=submissions_in_progress_status,
+    ).count()
+
+    if (
+        submissions_in_progress
+        >= challenge_phase.max_concurrent_submissions_allowed
+    ):
+        message = "You have {} submissions that are being processed. \
+                   Please wait for them to finish and then try again."
+        response_data = {"error": message.format(submissions_in_progress)}
+        return Response(response_data, status=status.HTTP_406_NOT_ACCEPTABLE)
+
+    serializer = SubmissionSerializer(
+        data=request.data,
+        context={
+            "participant_team": participant_team,
+            "challenge_phase": challenge_phase,
+            "request": request,
+        },
+    )
+
+    submission_message = {
+        "challenge_pk": challenge_pk,
+        "phase_pk": challenge_phase_pk,
+    }
+
+    if serializer.is_valid():
+        serializer.save()
+        submission = serializer.instance
+
+        submission_message["submission_pk"] = submission.pk
+
+        file_ext = os.path.splitext(request.data["file_name"])[-1]
+        file_name = "submission_files/presigned_url_files/submission_{}/{}{}".format(
+            submission.pk, uuid.uuid4(), file_ext
+        )
+        file_key = "{}/{}".format(settings.MEDIAFILES_LOCATION, file_name)
+
+        response = get_presigned_url_for_file_upload(file_key)
+        if response.get("error"):
+            response_data = response
+            return Response(response_data, status=status.HTTP_400_BAD_REQUEST)
+
+        submission.input_file = file_name
+        submission.save()
+
+        response_data = {
+            "presigned_url": response.get("presigned_url"),
+            "submission_message": submission_message,
+        }
+        return Response(response_data, status=status.HTTP_201_CREATED)
+    return Response(serializer.errors, status=status.HTTP_406_NOT_ACCEPTABLE)
+
+
+@api_view(["POST"])
+@throttle_classes([UserRateThrottle])
+@permission_classes((permissions.IsAuthenticated, HasVerifiedEmail))
+@authentication_classes((ExpiringTokenAuthentication,))
+def publish_submission_message_api(request):
```

Comment:
Can we have all the checks of creating a submission by the participant here as well otherwise I can simply write a script to post vague messages into the queue?

Reply:
Yeah, thanks for pointing out. I'll make the change.